jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java

```diff
@@ -172,6 +175,11 @@ public ProcessInstance getProcessInstance(long id, boolean readOnly) {
 
             if (readOnly) {
                 internalRemoveProcessInstance(processInstance);
+                if (ppcm instanceof JpaProcessPersistenceContextManager) { // for MapBasedPersistenceTest
```

Comment:
In order to avoid the instanceof here, I would add a new default method to ProcessPersistenceContextManager with signature checkProcessInstanceInfo (ProcessInstanceInfo processInstanceInfo, boolean readOnly). 
The other alternative is to modify getProcessPersistenceContext signature adding boolean readOnly parameter and do the detach there (but this will imply that both this methods and getPRocessPersistenceContext has to be invoke findProcessInstanteInfo in the same thread)